### PR TITLE
DRAFT: Don't immediately renegotiate for initial streams

### PR DESCRIPTION
**Pull request by feross**
Fixes https://github.com/feross/simple-peer/issues/670

This is a draft because I don't want to change the public API in order to fix this bug.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -146,7 +146,7 @@ class Peer extends stream.Duplex {
 
     if (this.streams) {
       this.streams.forEach(stream => {
-        this.addStream(stream)
+        this.addStream(stream, false)
```
Comment (feross) [inline index.js:149]: This parameter tells `addTrack` that it should not trigger a renegotiation.

File: index.js [hunk 2]
```diff
@@ -270,11 +270,11 @@ class Peer extends stream.Duplex {
    * Add a MediaStream to the connection.
    * @param {MediaStream} stream
    */
-  addStream (stream) {
+  addStream (stream, triggerNegotiate = true) {
```
Comment (feross) [inline index.js:273]: I don't actually want to add this to the public API, but this PR is just to demonstrate a fix to the problem

File: index.js [hunk 3]
```diff
@@ -373,6 +373,9 @@ class Peer extends stream.Duplex {
   }
 
   negotiate () {
+    this._debug('negotiate')
+    this.emit('negotiate')
```
Comment (feross) [inline index.js:377]: Unrelated fix here: I moved this debug log and the undocumented `'negotiate'` event to fire when negotiate happens. Before, it was firing when the connection becomes stable, which seemed like a bug

File: index.js [hunk 4]
```diff
@@ -882,6 +885,7 @@ class Peer extends stream.Duplex {
     if (this.destroyed) return
 
     if (this._pc.signalingState === 'stable' && !this._firstStable) {
+      this._firstStable = false
```
Comment (feross) [inline index.js:888]: This boolean seemed to be set in the wrong place. Before it was being set after any signalingStateChange instead of after the first stable state. I moved it inside the if statement so it only fires after first stable.

**Comment by daxfrost** [comment]
Just tested this, it's working great! Looking forward to a release with this in.

**Comment by t-mullen** [comment]
I've implemented a fix for this in #694, with tests and more debugging logs.

The causes were multiple:
- #570 changed the initial value of `_isNegotiating` to use the `channelConfig.negotiated` option (whether the datachannel is negotiated isn't relevant to the negotiation of the peer).
- `_firstStable` was set in the wrong place.
- The entire approach of dropping the non-initiator's 1st queued renegotiation request on first stable is wrong - this may be a valid request if we add a track right after creating an answer. What we actually want to do is start negotiation on both peers and immediately discard the non-initiator's 1st batched negotiation (since we know the initiator will negotiate). This was an unlikely race condition, but I've fixed it anyways.

File: index.js (hunk 3 above)
Comment (t-mullen) [reply index.js:378]: It's meant to fire when negotiation completes, which coincides with `stable` without a queued renegotiation.

It's only used in some tests.

Comment (feross) [reply index.js:378]: I think it'd be better to name the event something other than `'negotiate'` since the convention is to emit debug logs and events named after methods when they're first called. Maybe it should be called `'negotiated'` then?

**Comment by feross** [comment]
Superceded by #722
